### ISSUE TITLE
Enabler: Create a Field for Storing and Validating Data

### ISSUE THREAD

**Issue by bryanpizzillo** [NCIOCPL/cgov-digital-platform#3337]
# Create a JSON Field for Storing and Validating Data
> **ESTIMATE** TBD

As part of changing out the Mega Menu for NCIDS, we no longer want users to edit the raw HTML. There have been issues in the past where extra spaces were left in, or the HTML was malformed, and this leads to a broken site. We would like to use structured data to define the items in the megamenu, this data will be consumed by the front-end code as JSON. We will treat the Mega Menu Content as a `block_content` entity so that we can leverage all the management interfaces, translation and workflow without having to roll our own as we would in an Entity. In a perfect word we would have client-side tool that allows users to layout the mega menu and select the links for it -- however that is out of scope for now, and might never be worth the LOE. So we can iteratively build up this functionality though. In discussions with IAs, manually editing JSON will be problematic as there is a high potential for errors. YAML would provide a more human-friendly version of JSON. Also as YAML can be converted to JSON and vice-versa it will work in our backend. Finally we can define a JSONSchema that would be used to validate the YAML. In a more perfect world, we could use a JavaScript code editor (e.g., Ace, Monoco, etc) to give a nicer user interface and error highlighting. However, this friendlier UI could be an add on.

A review of third-party Drupal modules shows there is nothing that is maintained that we can leverage. `json_field` looked closest, but has no validation and no YAML editing. By the time we recreate the widgets, formatters and constraints, we might as well just make our own field.

### Requirements
1. This should be placed in a `json_data_field` module in `docroot/modules/custom/json_data_field`.
2. There should be tests created. See the [app_modules](https://github.com/NCIOCPL/cgov-digital-platform/blob/develop/docroot/modules/custom/app_module/tests/src/Functional/AppModuleFieldBrowserTestBase.php) tests for examples.
3. The goal is to allow editors to edit a block of data using YAML, validate it against a JSONSchema on save, and get the data out of the field as JSON.

#### Field Plugin Components
* **Field Type**
    * **Annotations** (`@FieldType`)
       * **id:** `json_data`
       * **label:** `JSON (stored as text in database)`
       * **description:** `Allows JSON data to be stored in the database. In the database the content is stored in a text column.`
       * **category:** `Data`
       * **default_widget:** `yaml_textarea`
       * **default_formatter:** `json_data_formatter`
       * **constraints:**
          * Max length: This should be around `4194304`.  This works out to a 16MB/utf8mb4 column size ((16*2^20)/4) **IF** using medium size for column.
          * Validate against JSONSchema. It looks like this constrain will need to be created. Use `composer require justinrainbow/json-schema` to add the validator. There is an [example of usage in the jsonapi module.](https://github.com/drupal/drupal/blob/515d10367bbe5cc158153a90e7960f92c2862745/core/modules/jsonapi/src/EventSubscriber/ResourceResponseValidator.php#L167)
    * **NOTES:**
       * This should extend `FieldItemBase`
       * Schema: We should make the type of this be JSON dataThe data type of the field should be text, `big` is probably too big (4gb), `medium` (16MB) is even probably overkill, but `normal` (64KB) might be too small. :
          ```
          $schema['columns']['value'] = [
            'description' => 'Json data.',
            'type' => 'text',
            'size' => 'medium',
            'serialize' => TRUE,
          ];
          ```
       * There should be a constraint setup for the max length.
       * We should also be able to provide a JSONscheme for validation.  
* **Field Widget**
   * _NOTE: This is where we can get creative in the future with fancier editors, but not now._
   * **Annotations** (`@FieldWidget`)
      * **id:** `yaml_textarea`
      * **label:** `Plain textarea (multiple rows)`
      * **field_types:** `json_data`
   * **NOTES:** 
      * Without digging in too deep it seems that this can extend `StringTextareaWidget`. YMMV.
      * We will need to translate from JSON -> YAML when displaying the field on edit
      * Translate from YAML -> JSON upon saving. 
      * The methods I have done this for in the [app_module](https://github.com/NCIOCPL/cgov-digital-platform/blob/develop/docroot/modules/custom/app_module/src/Plugin/Field/FieldWidget/AppModuleReferenceSelectWidget.php) before were: 
         * `formElement` - Sets up editing form
            * Turn data into YAML (See `\Symfony\Component\Yaml\Dumper::dump`)
            * You will need to set the dump options so the YAML is pretty enough for IAs.
            * add validateData callback
         * `validateData` - Validates the data
         * `massageFormValues` - called before storing in DB
            * turn YAML into Data. (See `\Symfony\Component\Yaml\Parser::parse`)
      * As the constraint for validating against a schema is defined above on the "Field Type" above there should not be anything special to handle validation on the widget.
* **Field Formatter**
   * **Annotations** (`@FieldFormatter`)
      * **id:** `json_data_formatter`
      * **label:** `Default formatter for json_data fields` 
      * **field_types:** json_data
   * **NOTES:** 
      * This should just spit out the YAML onto the page in a `<code>` tag. Really it will never be used as we are doing all this to be used by an API. 

### Additional Notes
I was torn between making the schema marked as serializable or not. 
* We can do the following fairly easily: `YAML -> PHP -> DB -> PHP -> JSON` There is no simple `YAML -> JSON` within core Drupal/Symphony. You have to go `YAML -> PHP -> JSON`.  There are built in functions to do `PHP -> JSON` and `PHP -> YAML`, so there is no reason to not keep things in the DB serialized data for conversion purposes.
* Our YAML content actually supports YAML/json with serialized data within the loading process without issue. ([Example of this](https://github.com/NCIOCPL/cgov-digital-platform/blob/develop/docroot/profiles/custom/cgov_site/modules/custom/cgov_yaml_content/content/009_app_module_glossary_genetics.content.yml#L40))
* The only hesitation with serialized data is that when doing a select in the Database we can't really read it. However, to be fair, using SQLC to view content in the database is fairly useless. 

### References
* [json_field](https://git.drupalcode.org/project/json_field/-/tree/8.x-1.x/) module - this has so nice Kernel test cases.
* Our [app_module](https://github.com/NCIOCPL/cgov-digital-platform/blob/develop/docroot/modules/custom/app_module) module. This actually stores JSON, albeit serialized, for the app module settings of a reference field. 

Related: https://github.com/NCIOCPL/ncids/issues/460 (YAML Content for Megamenu)

**Comment by VictoriaSunNIH** [comment]
Providing the YAML content for English and Spanish, and cross linking it here: https://github.com/NCIOCPL/ncids/issues/460

**Comment by bryanpizzillo** [comment]
Closed by #3397
